cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesTable.java

```diff
@@ -148,6 +148,143 @@ public MetricsScanner scan(MetricsScanQuery query) throws OperationException {
+    // returned from the scan.
+    do {
+      FuzzyRowFilter filter = getFilter(query, -1, -1, false);
+      ScannerFields fields = new ScannerFields(startRow, endRow, columns, filter);
+      Scanner scanner = null;
+      try {
+        scanner = timeSeriesTable.scan(fields.startRow, fields.endRow, fields.columns, fields.filter);
+      } catch (Exception e) {
+        throw new OperationException(StatusCode.INTERNAL_ERROR, e.getMessage(), e);
+      }
+
+      rowResult = scanner.next();
+      if (rowResult != null) {
+        byte[] rowKey = rowResult.getRow();
+        int offset = 0;
+        String contextStr = entityCodec.decode(MetricsEntityType.CONTEXT, rowKey, offset);
+
+        if (query.getContextPrefix() != null && !contextStr.startsWith(query.getContextPrefix())) {
+          // if retrieved rowkey's contextPrefix does not match with the contextPrefix in query,
+          // we stop scanning and return
+          scanner.close();
+          break;
+        }
+
+        offset += entityCodec.getEncodedSize(MetricsEntityType.CONTEXT);
+
+        if (isContextQuery) {
+          metricsScanResults.add(contextStr);
+          // With the next scan we fast-forward to the next row that has different context name
+          // that we are searching for.
+          startRow = getNextRow(rowKey, contextOffset, contextLength);
+          if (startRow == null) {
+            //reached max possible key for the context, we will stop scanning now and return.
+            scanner.close();
+            break;
+          }
+        } else {
+          String metricName = entityCodec.decode(MetricsEntityType.METRIC, rowKey, offset);
+          metricsScanResults.add(metricName);
+          // With the next scan we fast-forward to the next row that has different metric name part
+          // that we are searching for.
+          startRow = getNextRow(rowKey, metricsOffset, metricsLength);
+          if (startRow == null) {
+            //reached max possible metrics key for the given context, we will stop scanning now and return.
+            scanner.close();
+            break;
+          }
+        }
+      }
+      // no more matching rows found
+      scanner.close();
+    } while (rowResult != null);
+
+    return metricsScanResults;
+  }
+
+  /*
+   * Returns row key that is:
+   * greater than given rowKey AND
+   * has same bytes in the beginning (up to the given offset) AND
+   * has different bytes in the range [offset..offset+length]
+   * Returns null if no row key with this properties exist.
+   */
+  private byte[] getNextRow(byte[] rowKey, int offset, int length) {
+    byte[] stopKey = Bytes.stopKeyForPrefix(Arrays.copyOfRange(rowKey,
+                                                               offset, offset + length));
+    if (stopKey == null) {
+      return null;
+    }
+    // next row-key
+    return  Bytes.concat(Bytes.head(rowKey, offset), stopKey,
```

Comment:
@chtyim we are making an RPC call which is more expensive , so we decided to go with this.